GetStream/stream-chat-android
# [4609] Expose the option to show the moderated message dialog

### 🎯 Goal

closes #4609 

### 🛠 Implementation details

- Moved the default content of `MessageListView.moderatedMessageLongClickListener` to a separate publicly visible function `MessageListView.showModeratedMessageDialog()`

### 🧪 Testing

Test A -> Verifying old functionality still works:

1. Launch the XML app
2. Send a message that will fail moderation
3. Long click on the moderated message

Expectation: The moderated message should appear and you should be able to exercise all options normally.

Test B -> Verifying that users override the default message long click listener while preserving all default SDK functionality.


<details>

<summary>Provide the patch summary here</summary>

```
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt	(revision 4e70b787c11aab2a429eaf5de8c35627e159bb7f)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt	(date 1674654136144)
@@ -17,6 +17,7 @@
 package io.getstream.chat.ui.sample.feature.chat
 
 import android.os.Bundle
+import android.util.Log
 import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
@@ -27,8 +28,10 @@
 import androidx.navigation.fragment.findNavController
 import androidx.navigation.fragment.navArgs
 import com.getstream.sdk.chat.utils.extensions.getCreatedAtOrThrow
+import com.getstream.sdk.chat.utils.extensions.isModerationFailed
 import com.getstream.sdk.chat.viewmodel.MessageInputViewModel
 import com.getstream.sdk.chat.viewmodel.messages.MessageListViewModel
+import io.getstream.chat.android.client.ChatClient
 import io.getstream.chat.android.client.errors.ChatNetworkError
 import io.getstream.chat.android.client.models.Flag
 import io.getstream.chat.android.client.models.Message
@@ -44,11 +47,15 @@
 import io.getstream.chat.android.core.ExperimentalStreamChatApi
 import io.getstream.chat.android.core.internal.InternalStreamChatApi
 import io.getstream.chat.android.livedata.utils.EventObserver
+import io.getstream.chat.android.ui.ChatUI
+import io.getstream.chat.android.ui.common.extensions.isGiphyNotEphemeral
+import io.getstream.chat.android.ui.common.extensions.isInThread
 import io.getstream.chat.android.ui.message.composer.viewmodel.MessageComposerViewModel
 import io.getstream.chat.android.ui.message.composer.viewmodel.bindView
 import io.getstream.chat.android.ui.message.input.viewmodel.bindView
 import io.getstream.chat.android.ui.message.list.header.viewmodel.MessageListHeaderViewModel
 import io.getstream.chat.android.ui.message.list.header.viewmodel.bindView
+import io.getstream.chat.android.ui.message.list.options.message.MessageOptionItemsFactory
 import io.getstream.chat.android.ui.message.list.viewmodel.bindView
 import io.getstream.chat.android.ui.message.list.viewmodel.factory.MessageListViewModelFactory
 import io.getstream.chat.ui.sample.common.navigateSafely
@@ -93,6 +100,7 @@
         initChatViewModel()
         initMessagesViewModel()
         configureBackButtonHandling()
+        overrideDefaultLongClickListener()
 
         if (ToggleService.isEnabled(ToggleService.TOGGLE_KEY_MESSAGE_COMPOSER)) {
             initMessageComposerViewModel()
@@ -297,4 +305,32 @@
         val chatError = error() as ChatNetworkError
         return chatError.streamCode == 4
     }
+
+    @OptIn(InternalStreamChatApi::class)
+    private fun overrideDefaultLongClickListener(){
+        binding.messageListView.setMessageLongClickListener { message ->
+            Log.d("CustomLongClickListener", "Firing!")
+            childFragmentManager.let { fragmentManager ->
+                if (message.isModerationFailed(ChatClient.instance())) {
+                    binding.messageListView.showModeratedMessageDialog(message)
+                } else {
+                    val style = binding.messageListView.requireStyle()
+                    val isEditEnabled = style.editMessageEnabled && !message.isGiphyNotEphemeral()
+                    val viewStyle = style.copy(editMessageEnabled = isEditEnabled)
+
+                    val messageOptionItemsFactory = MessageOptionItemsFactory.defaultFactory(requireContext())
+
+                    val messageOptionItems = messageOptionItemsFactory.createMessageOptionItems(
+                        selectedMessage = message,
+                        currentUser = ChatUI.currentUserProvider.getCurrentUser(),
+                        isInThread = messageListViewModel.mode.value is MessageListViewModel.Mode.Thread || message.isInThread(),
+                        ownCapabilities = messageListViewModel.ownCapabilities.value ?: setOf(),
+                        style = viewStyle
+                    )
+
+                    binding.messageListView.showMessageOptionsDialog(fragmentManager, message, messageOptionItems)
+                }
+            }
+        }
+    }
 }
```

</details>


1. Apply the patch above
2. Re-run the steps from Test A and verify that the behavior is identical
3. Enter a thread and long click a non-moderated message and verify that the message options have been adjusted accordingly (You shouldn't be able to thread reply in a thread for instance)

Expectation: No behavioral changes should occur compared to Test A, you should also see a logcat message with the tag `CustomLongClickListener` in the logs.  


### ☑️Contributor Checklist

#### General

#### Code & documentation

### ☑️Reviewer Checklist

### 🎉 GIF

![gif](https://media3.giphy.com/media/hhgtaQAWWLO7RyG171/giphy.gif?cid=ecf05e47ouktdc3xlcywij5gs9o6o4bevw85qdnedj7jdyyc&rid=giphy.gif&ct=g)
